casper/src/main/scala/coop/rchain/casper/BlockDag.scala

```diff
@@ -6,10 +6,26 @@ import coop.rchain.casper.protocol.BlockMessage
 
 import scala.collection.immutable.{HashMap, HashSet}
 
+/*
+ * toposort - Topological sort of the DAG. The index in the outer Vector
+ *            is the block number (length of the longest path back to
+ *            genesis) and the index of the inner Vector breaks ties for
+ *            blocks with the same number based when the node observed the block.
+ *            `topoSort.flatten` gives the sorted blocks.
```

Comment:
"based on when the node observed"

Does this not make it non-deterministic for validators?

Reply:
Yes, each node could have a different sorting. But they would only differ for blocks at the same height (i.e. blocks with no causal relation), which does not matter. Moreover, this is the only way to ensure that the sorting is stable under append, which I think was a requirement for easy update of the flattened version.

Reply:
We could use something like validator ID, but then we have to do some work when deciding where to insert a new block in the sorting (where as right now it is a very minimal operation). I don't see why it matters what order they are in for the sorting though. If there was indeed a race between two blocks at the same height then they conflict and so could never be merged together anyways.

Reply:
As of now; that's why I said eventual implementation.

Reply:
In the eventual implementation the merge preference is determined by the block's score, so the sort still won't matter. As far as I can tell, the sorting is an internal detail which allows us to compute the uncommon ancestors and serialize the DAG to disk, but different nodes don't need to agree on the precise ordering because any topological sort will do for those purposes.

Reply:
@birchmd Exactly. Any topological sort will do for the algorithm. You'll get the same results with any.